packages/ember-application/tests/system/engine_instance_initializers_test.js

```diff
@@ -0,0 +1,366 @@
+  let firstInitializerRunCount = 0;
+  let secondInitializerRunCount = 0;
+  let FirstEngine = Engine.extend();
+  let firstEngine, firstEngineInstance;
+
+  FirstEngine.instanceInitializer({
+    name: 'first',
+    initialize(engine) {
+      firstInitializerRunCount++;
+    }
+  });
+
+  let SecondEngine = Engine.extend();
+  let secondEngine, secondEngineInstance;
+
+  SecondEngine.instanceInitializer({
+    name: 'second',
+    initialize(engine) {
+      secondInitializerRunCount++;
+    }
+  });
+
+  firstEngine = FirstEngine.create();
+  firstEngineInstance = firstEngine.buildInstance();
+
+  return firstEngineInstance.boot()
+    .then(() => {
+      equal(firstInitializerRunCount, 1, 'first initializer only was run');
+      equal(secondInitializerRunCount, 0, 'first initializer only was run');
+
+      secondEngine = SecondEngine.create();
+      secondEngineInstance = secondEngine.buildInstance();
+      return secondEngineInstance.boot();
+    })
+    .then(() => {
+      equal(firstInitializerRunCount, 1, 'second initializer only was run');
+      equal(secondInitializerRunCount, 1, 'second initializer only was run');
+
+      run(function() {
+        firstEngineInstance.destroy();
+        secondEngineInstance.destroy();
+
+        firstEngine.destroy();
+        secondEngine.destroy();
+      });
+    });
+});
+
+QUnit.test('initializers are concatenated', function() {
+  let firstInitializerRunCount = 0;
+  let secondInitializerRunCount = 0;
+  let FirstEngine = Engine.extend();
+
+  FirstEngine.instanceInitializer({
+    name: 'first',
+    initialize(engine) {
+      firstInitializerRunCount++;
+    }
+  });
+
+  let SecondEngine = FirstEngine.extend();
+
+  SecondEngine.instanceInitializer({
+    name: 'second',
+    initialize(engine) {
+      secondInitializerRunCount++;
+    }
+  });
+
+  let firstEngine = FirstEngine.create();
+  let firstEngineInstance = firstEngine.buildInstance();
+
+  let secondEngine, secondEngineInstance;
+
+  return firstEngineInstance.boot()
+    .then(() => {
+      equal(firstInitializerRunCount, 1, 'first initializer only was run when base class created');
+      equal(secondInitializerRunCount, 0, 'second initializer was not run when first base class created');
+      firstInitializerRunCount = 0;
+
+      secondEngine = SecondEngine.create();
+      secondEngineInstance = secondEngine.buildInstance();
+      return secondEngineInstance.boot();
+    })
+    .then(() => {
+      equal(firstInitializerRunCount, 1, 'first initializer was run when subclass created');
+      equal(secondInitializerRunCount, 1, 'second initializers was run when subclass created');
+
+      run(function() {
+        firstEngineInstance.destroy();
+        secondEngineInstance.destroy();
+
+        firstEngine.destroy();
+        secondEngine.destroy();
+      });
+    });
+});
+
+QUnit.test('initializers are per-engine', function() {
+  expect(0);
```

Comment:
I'll add some verification that initializers raise if duplicated, which would make both tests more rigorous.